### PR TITLE
Distributed Runtime

### LINKED ISSUE #120: [FEA] Add convenient methods to EncodedObject
EncodedObject should add new methods for common patterns.

This should either update the behavior of `add_memory_block` or provide a new method that does the following:

- takes a buffer_view `(addr, size, kind)
- checks to see if the `addr` is registered with the data plane
- if so, make a remote memory descriptor from the view,
- if not, allocate a buffer, copy the data to the buffer and make a remote md to the newly allocated buffer.
- optionally, we could decided to registered the unregistered buffer if the size is sufficiently large to avoid a copy.

It should be noted, that this method should only be called for creating a contiguous block of memory in the encoded object of a sufficiently large size to warrant a remote memory descriptor.

For smaller objects, we might find we should use a ReusablePool of buffers and used packed descriptors into those buffers.  The reusable pool offers an optimization over just allocating a buffer from the registered memory resource.

### PR DESCRIPTION
**Pull request by ryanolson**
This PR is a major step toward the completion Multi-Node or fully distributed SRF pipelines.

This PR puts in place the framework for the following components:
- control plane
- data plane
- fiber based grpc
- ucx runtime

This PR improves upon the following existing components:
- codable protocol

This PR add the initial public runtime and resources interface.

Closes #120

### CODE REVIEW

File: include/srf/pubsub/subscriber.hpp
Comment (mdemoret-nv) [inline include/srf/pubsub/subscriber.hpp:53]: If this inherits from `public node::Queue<T>` that means you can make connections that write to the subscriber. i.e. `node::make_edge(source, subscriber)`. This needs to inherit from something else I think.

File: include/srf/runtime/api.hpp
Comment (mdemoret-nv) [inline include/srf/runtime/api.hpp:29]: These should be in "forward.hpp"

File: include/srf/runtime/api.hpp
Comment (mdemoret-nv) [inline include/srf/runtime/api.hpp:126]: This relationship seems weird to me. I feel like the public runtime objects shouldnt have friends. Let alone ones in a completely different namespace. All of the methods on these objects should be part of the public API.

File: include/srf/types.hpp
Comment (mdemoret-nv) [inline include/srf/types.hpp:44]: I added TagID here which helped distinguish it from InstanceID.

File: protos/CMakeLists.txt
Comment (mdemoret-nv) [inline protos/CMakeLists.txt:63]: All of the indentation here got messed up.

File: python/srf/_pysrf/include/pysrf/plugins.hpp
Comment (mdemoret-nv) [inline python/srf/_pysrf/include/pysrf/plugins.hpp:42]: Merge error

File: src/internal/pubsub/publisher.hpp
Comment (mdemoret-nv) [inline src/internal/pubsub/publisher.hpp:42]: Can we call this `PublisherService` to distinguish it from the public `Publisher`?

File: src/internal/pubsub/subscriber.hpp
Comment (mdemoret-nv) [inline src/internal/pubsub/subscriber.hpp:43]: Also call this `SubscriberService`

File: include/srf/pubsub/publisher.hpp
Comment (mdemoret-nv) [inline include/srf/pubsub/publisher.hpp:80]: I dont love this. Why is the `m_persistent_channel` needed here?

**Review by mdemoret-nv — APPROVED** [review]
Approving with comments.

File: include/srf/codable/api.hpp
Comment (ryanolson) [reply include/srf/codable/api.hpp:43]: I think we could do that as long as we properly annotate as private with IWYU. 

I've been using forward.hpp as the thing you use from another directory, and using explicit forward declares for the files in the directory. 

If you include forward.hpp in the definition header, I think IWYU makes you include that everywhere even though you don't need it.

File: include/srf/concepts/invocable.hpp
Comment (ryanolson) [reply include/srf/concepts/invocable.hpp:29]: Sneaky. Yeah, it's not used it got included accidentally. That should be in the cpp20 branch

File: include/srf/memory/buffer_view.hpp
Comment (ryanolson) [reply include/srf/memory/buffer_view.hpp:50]: Stylistically I agree, but fundamental types don't have move constructors. Moves degrade to copies in that case.

File: include/srf/memory/codable/buffer.hpp
Comment (ryanolson) [reply include/srf/memory/codable/buffer.hpp:32]: Yeah, that was just a carry over from using a is base of...

File: include/srf/node/operators/operator.hpp
Comment (ryanolson) [reply include/srf/node/operators/operator.hpp:79]: Good spot. That was where I was getting an exception throw when inheriting multiple level of enable shared from this.

File: include/srf/node/operators/unique_operator.hpp
Comment (ryanolson) [reply include/srf/node/operators/unique_operator.hpp:26]: Nor I. It's just a place holder until we get the new operator semantics. 

It's an operator that doesn't inherit from enable shared from this.

File: include/srf/pubsub/subscriber.hpp
Comment (ryanolson) [reply include/srf/pubsub/subscriber.hpp:53]: The new operator semantic should give us the option to make this a channel provider and not appear to have a dangling upstream connection. 

In fact, the upstrsm of the queue is connected to the internal pipeline that provides either a T or a RemoteDescriptkr.

File: include/srf/pubsub/publisher.hpp
Comment (ryanolson) [reply include/srf/pubsub/publisher.hpp:80]: Publisher is both directly writable and upstream connectable. 

Because of that, we can not just immediate request a control plane drop service on this tag. 

We have to wait to make the control plane drop request until after all upstream have been disconnected AND a call to request stop have been made on the object. 

We use are using the persistent channel model to do this. Initialize the persistent channel.  Upstream ingress acceptors can connect. Request stop drops the persistent connection. When the last connection is released the control plane messsge is issued.

File: protos/CMakeLists.txt
Comment (ryanolson) [reply protos/CMakeLists.txt:63]: Where is that formatting triggered? Am I missing an extension?

File: include/srf/runtime/api.hpp
Comment (ryanolson) [reply include/srf/runtime/api.hpp:126]: It's primary to limit the things that are public on the IPartition object. 

make_xxx_service are not really meant to be publicly callable, but need to be called by the create method on the publisher/subscriber. 

If there is a better way, im open to suggestions

File: src/internal/pubsub/publisher.hpp
Comment (ryanolson) [reply src/internal/pubsub/publisher.hpp:42]: Done

File: src/internal/pubsub/subscriber.hpp
Comment (ryanolson) [reply src/internal/pubsub/subscriber.hpp:43]: Done

File: include/srf/concepts/invocable.hpp
Comment (ryanolson) [reply include/srf/concepts/invocable.hpp]: Removed

File: python/srf/_pysrf/include/pysrf/plugins.hpp
Comment (ryanolson) [reply python/srf/_pysrf/include/pysrf/plugins.hpp]: Fixed.

File: include/srf/types.hpp
Comment (ryanolson) [reply include/srf/types.hpp]: Done

File: include/srf/node/operators/operator.hpp
Comment (ryanolson) [reply include/srf/node/operators/operator.hpp]: fixed

File: include/srf/node/operators/broadcast.hpp
Comment (ryanolson) [reply include/srf/node/operators/broadcast.hpp:79]: fixed

File: include/srf/memory/codable/buffer.hpp
Comment (ryanolson) [reply include/srf/memory/codable/buffer.hpp]: Fixed

File: cmake/dependencies.cmake
Comment (mdemoret-nv) [reply cmake/dependencies.cmake:123]: Ok, lets figure this out in the rename/refactor

File: include/srf/codable/api.hpp
Comment (mdemoret-nv) [reply include/srf/codable/api.hpp:43]: >If you include forward.hpp in the definition header, I think IWYU makes you include that everywhere even though you don't need it.

Interesting. Ok with these for now then.

File: include/srf/memory/buffer_view.hpp
Comment (mdemoret-nv) [reply include/srf/memory/buffer_view.hpp:54]: Fair enough. Then this needs to function more like the assignment operator and use `std::swap()` so that `other` is left in an uninitialized state.

File: protos/CMakeLists.txt
Comment (mdemoret-nv) [reply protos/CMakeLists.txt:63]: We will sort out the format settings in another PR

File: include/srf/pubsub/publisher.hpp
Comment (mdemoret-nv) [reply include/srf/pubsub/publisher.hpp:80]: After discussion, I'm fine with this for now.

File: include/srf/node/operators/unique_operator.hpp
Comment (mdemoret-nv) [reply include/srf/node/operators/unique_operator.hpp:26]: We can address after the non-linear pipeline stuff is merged.

File: include/srf/pubsub/subscriber.hpp
Comment (mdemoret-nv) [reply include/srf/pubsub/subscriber.hpp:53]: We can address after the non-linear pipeline stuff is merged.

File: include/srf/runtime/api.hpp
Comment (mdemoret-nv) [reply include/srf/runtime/api.hpp:126]: This is fine for now.

File: include/srf/runtime/api.hpp
Comment (mdemoret-nv) [reply include/srf/runtime/api.hpp:29]: We can address after IWYU changes are merged.

**Comment by mdemoret-nv** [comment]
@gpucibot merge

**Pull request merged**
